File: api/server-common/src/main/java/su/plo/voice/api/server/config/ServerLanguages.java
package su.plo.voice.api.server.config;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import su.plo.lib.api.chat.MinecraftTextComponent;
import su.plo.lib.api.chat.MinecraftTranslatableText;
import su.plo.lib.api.server.command.MinecraftChatHolder;

import java.io.File;
import java.util.Map;

public interface ServerLanguages {

    /**
     * Registers a new language using resources
     *
     * <p>
     *     Reads <b>languages/list</b> from resources using {@link ResourceLoader}
     *     <br/>
     *     In the <b>list</b> you should specify list of languages separated by <b>\n</b>
     *     <br/>
     *     After reading the list, all languages from it will be read from resources <b>languages/*.toml</b>
     *     and saved to <b>languageFolder/*.toml</b>
     * </p>
     *
     * <p>
     *     You can edit and create new languages in languagesFolder, they not will be overwritten or deleted
     * </p>
     *
     * <p>Default language is <b>en_us</b>, can be changed in server config</p>
     */
    void register(@NotNull ResourceLoader resourceLoader,
                  @NotNull File languagesFolder);

    /**
     * Registers a new language using crowdin
     *
     * <p>Works as {@link #register(ResourceLoader, File)}, but also uses crowdin as defauls</p>
     *
     * <p>Crowdin translations will be cached in <b>languageFolder/.crowdin</b> for <b>3 days</b></p>
     *
     * <p>Default language is <b>en_us</b>, can be changed in server config</p>
     */
    void register(@NotNull String crowdinProjectId,
                  @Nullable String fileName,
                  @NotNull ResourceLoader resourceLoader,
                  @NotNull File languagesFolder);

    /**
     * Gets the server language by name or default language if not found
     */
    Map<String, String> getServerLanguage(@Nullable String languageName);

    /**
     * Gets the client language by name or default language if not found
     */
    Map<String, String> getClientLanguage(@Nullable String languageName);

    /**
     * Gets the default server language
     */
    default Map<String, String> getServerLanguage() {
        return getServerLanguage((String) null);
    }

    /**
     * Gets the server language by chat holder
     */
    default Map<String, String> getServerLanguage(@NotNull MinecraftChatHolder holder) {
        return getServerLanguage(holder.getLanguage());
    }

    /**
     * Gets the default client language
     */
    default Map<String, String> getClientLanguage() {
        return getClientLanguage((String) null);
    }

    /**
     * Gets the client language by chat holder
     */
    default Map<String, String> getClientLanguage(@NotNull MinecraftChatHolder holder) {
        return getClientLanguage(holder.getLanguage());
    }

    /**
     * Translates text using server language
     */
    default MinecraftTextComponent translate(@NotNull MinecraftTranslatableText text,
                                             @NotNull MinecraftChatHolder holder,
                                             @NotNull String key) {
        Map<String, String> language = getServerLanguage(holder.getLanguage());
        if (!language.containsKey(key)) return text;

        return MinecraftTextComponent.literal(language.get(key));
    }
}

File: api/server-common/src/main/kotlin/su/plo/voice/api/server/language/ServerLanguages.kt
package su.plo.voice.api.server.language

import su.plo.slib.api.chat.component.McTextComponent
import su.plo.slib.api.chat.component.McTranslatableText
import su.plo.slib.api.command.McChatHolder
import su.plo.slib.api.resource.ResourceLoader
import java.io.File
import java.util.concurrent.CompletableFuture

/**
 * Manages languages.
 */
interface ServerLanguages {

    /**
     * Gets or sets whether crowdin is enabled.
     *
     * `true` by default.
     *
     * @return `true` if crowdin is enabled, `false` otherwise.
     */
    var crowdinEnabled: Boolean

    /**
     * Registers a new language using resources.
     *
     * Reads the language list from resources using [ResourceLoader]. The list should specify the available
     * languages separated by '\n'. After reading the list, all languages from it will be read from resources
     * as 'languages/?.toml' and saved to 'languagesFolder/?.toml'. Languages created or edited in the
     * 'languagesFolder' will not be overwritten or deleted.
     *
     * Default language is 'en_us'
     *
     * @param resourceLoader   The resource loader for reading language resources.
     * @param languagesFolder  The folder where language files will be saved.
     * @return A CompletableFuture representing the completion of the registration process.
     */
    fun register(
        resourceLoader: ResourceLoader,
        languagesFolder: File
    ): CompletableFuture<Void>

    /**
     * Registers a new language using crowdin.
     *
     * This method works similarly to [register], but it also uses crowdin as the default language source.
     * Crowdin translations will be cached in 'languageFolder/.crowdin' for 3 days.
     *
     * @param crowdinProjectId The crowdin project ID for translations.
     * @param fileName         The name of the language file in crowdin (null for default).
     * @param resourceLoader   The resource loader for reading language resources.
     * @param languagesFolder  The folder where language files will be saved.
     * @return A CompletableFuture representing the completion of the registration process.
     */
    fun register(
        crowdinProjectId: String,
        fileName: String?,
        resourceLoader: ResourceLoader,
        languagesFolder: File
    ): CompletableFuture<Void?>

    /**
     * Gets server language data by name or default language if not found.
     *
     * @param languageName The name of the language to retrieve.
     * @return A map representing the server language data.
     */
    fun getServerLanguage(languageName: String?): Map<String, String>

    /**
     * Gets client language data by name or default language if not found.
     *
     * @param languageName The name of the language to retrieve.
     * @return A map representing the client language data.
     */
    fun getClientLanguage(languageName: String?): Map<String, String>

    /**
     * Gets the default server language.
     *
     * @return A map representing the default server language data.
     */
    val serverLanguage: Map<String, String>
        get() = getServerLanguage(null)

    /**
     * Gets server language data based on the chat holder's language setting.
     *
     * @param holder The chat holder whose language setting is used.
     * @return A map representing the server language data.
     */
    fun getServerLanguage(holder: McChatHolder): Map<String, String> =
        getServerLanguage(holder.language)

    /**
     * Gets the default client language.
     *
     * @return A map representing the default client language data.
     */
    val clientLanguage: Map<String, String>
        get() = getClientLanguage(null)

    /**
     * Gets client language data based on the chat holder's language setting.
     *
     * @param holder The chat holder whose language setting is used.
     * @return A map representing the client language data.
     */
    fun getClientLanguage(holder: McChatHolder): Map<String, String> {
        return getClientLanguage(holder.language)
    }

    /**
     * Translates text using the server language.
     *
     * @param text   The text to be translated.
     * @param holder The chat holder whose language setting is used.
     * @param key    The translation key.
     * @return A [McTextComponent] representing the translated text.
     */
    fun translate(
        text: McTranslatableText,
        holder: McChatHolder,
        key: String
    ): McTextComponent? {
        val language = getServerLanguage(holder.language)
        return if (!language.containsKey(key)) text else McTextComponent.literal(language[key]!!)
    }
}
